ui/src/app/views/interactive-notebook.tsx

```diff
@@ -0,0 +1,175 @@
+    borderRight: '1px solid ' + colors.gray[5],
+    backgroundColor: 'rgba(38,34,98,0.05)',
+    textAlign: 'center',
+    lineHeight: '40px'
+  },
+  active: {
+    backgroundColor: 'rgba(38,34,98,0.2)'
+  },
+  clickable: {
+    cursor: 'pointer'
+  },
+  disabled: {
+    cursor: 'not-allowed',
+    opacity: 0.7
+  },
+  navBarIcon: {
+    height: '16px',
+    width: '16px',
+    verticalAlign: 'middle',
+    marginLeft: 0,
+    marginRight: '5px',
+    marginBottom: '3px'
+  },
+  previewFrame: {
+    width: '100%',
+    height: 800,
+    border: 0
+  },
+  rotate: {
+    animation: 'rotation 2s infinite linear'
+  }
+});
+
+interface Props {
+  workspace: WorkspaceData;
+}
+
+interface State {
+  clusterStatus: ClusterStatus;
+  userRequestedEditMode: boolean;
+  html: string;
+}
+
+export const InteractiveNotebook = withCurrentWorkspace()(
+  class extends React.Component<Props, State> {
+
+    private billingProjectId = urlParamsStore.getValue().ns;
+    private workspaceId = urlParamsStore.getValue().wsid;
+    private nbName = urlParamsStore.getValue().nbName;
+
+    constructor(props) {
+      super(props);
+      this.state = {
+        userRequestedEditMode: false,
+        clusterStatus: ClusterStatus.Unknown,
+        html: ''
+      };
+    }
+
+    componentDidMount(): void {
+      workspacesApi().readOnlyNotebook(this.billingProjectId, this.workspaceId, this.nbName)
+        .then(html => {
+          this.setState({html: html.html});
+        });
+    }
+
+    private runCluster(): void {
+      const retry = () => {
+        setTimeout(() => this.runCluster(), 5000);
+      };
+
+      clusterApi().listClusters(this.billingProjectId)
+        .then((body) => {
+          const cluster = body.defaultCluster;
+          this.setState({clusterStatus: cluster.status});
+
+          if (cluster.status === ClusterStatus.Stopped) {
+            notebooksClusterApi()
+              .startCluster(cluster.clusterNamespace, cluster.clusterName);
+          }
+
+          if (cluster.status === ClusterStatus.Running) {
+            this.onClusterRunning();
+          } else {
+            retry();
+          }
+        })
+        .catch(() => {
+          retry();
+        });
+    }
+
+    private onEditClick() {
+      if (this.canWrite()) {
+        this.setState({userRequestedEditMode: true});
+        this.runCluster();
+      }
+    }
+
+    private onClusterRunning() {
+      navigate(['workspaces', this.billingProjectId, this.workspaceId, 'notebooks', this.nbName]);
+    }
+
+    private canWrite() {
+      return WorkspacePermissionsUtil.canWrite(this.props.workspace.accessLevel);
+    }
+
+    render() {
+      return (
+        <div>
+          <div style={styles.navBar}>
+            <div style={{...styles.navBarItem, ...styles.active, width: 227}}>
+              Preview (Read-Only)
+            </div>
+            {this.state.userRequestedEditMode ? (
+              <div style={{...styles.navBarItem, width: 620}}>
```

Comment:
620 seems like a large value to hardcode, where'd this come from? In the mocks the buttons are `229`, the loading dialog is 509px. Might be better to let these flow more naturally though, e.g. `display: flex`

Reply:
tried using flex and the styling is SO MUCH better now. Thanks